### DISCUSSION TITLE
Console State Integration

### DISCUSSION THREAD

**Discussion by MrOwnership** [RetroAchievements/RAWeb#1065] [Ideas]
I've started to look into adding a `State` field to the `Console` database table and allowing it to be updated by admins. I figured it's worth some discussion before diving to far into implementation.

My thoughts on the new states that would be useful to have.
| State               | Definition                                                                           |
| :------------------ | :----------------------------------------------------------------------------------- |
| Inactive            | No current work to support RA being done.                                            |
| Initial Development | Core or Emulator compatibility in development. RA compatibility testing in progress. |
| Rollout             | Core has been tested and working properly. Set development approved and in progress. |
| Validated           | Console is validated and achievements can be earned by the community.                |

With added states we can add in restrictions on what can and cannot be done for console in those states. I'm not heavily involved in the `Initial Development` or `Rollout` phases on consoles so there is likely some aspects that I've missed, but here are my initial thoughts on permissions based off the new states.

|                         |               Inactive               | Initial Development |      Rollout       |     Validated      |
| ----------------------- | :----------------------------------: | :-----------------: | :----------------: | :----------------: |
| Game Page Comments      |          :heavy_check_mark:          | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |
| Set Request             |          :heavy_check_mark:          | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |
| Forum Topic Creation    |                                      |                     | :heavy_check_mark: | :heavy_check_mark: |
| Game Info Updates       |          :heavy_check_mark:          | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |
| Emu Recognizes ROM      |                                      | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |
| Code Notes              |                                      | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |
| Rich Presence Update    |                                      |                     | :heavy_check_mark: | :heavy_check_mark: |
| Leaderboard Submissions |                                      |                     |                    | :heavy_check_mark: |
| Leaderboard Uploads     |                                      |                     | :heavy_check_mark: | :heavy_check_mark: |
| Unofficial Achievement  |                                      |                     | :heavy_check_mark: | :heavy_check_mark: |
| Core Achievements       |                                      |                     |                    | :heavy_check_mark: |
| Set Claim               |                                      |                     | :heavy_check_mark: | :heavy_check_mark: |
| Link Hash (Dev)         |                                      |                     | :heavy_check_mark: | :heavy_check_mark: |
| Link Hash (Admin)       | :heavy_check_mark: (Hubs and Events) | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |

Would there be any reason to prevent game page comments or set requests for consoles that are inactive?

There probably should not be any set development related work done until the core or emulator is confirmed to be working properly with RA, in which case the state of the console can be moved to `Rollout`.

A quick proof of concept admin tool for managing console states, which would also have the ability to add new consoles. Updating the ID and name of existing consoles would be disabled, that is not something that can be easily done as those IDs and names are used in more than just the web server. With this implementation we can validate new consoles without the need to have a PR merged in and deployed.
![](https://user-images.githubusercontent.com/16140035/180839701-b348d304-f973-4b9b-9bb8-3af4b8b825d9.gif)

**Comment by Jamiras** [comment]
I'm hesitant to include Code Notes in the Initial Development phase. The memory map may still be in flux at that point.

Leaderboard Uploads is also kind of up in the air. I think there's a loophole right now where you can create leaderboards for non-validated consoles through the website, but not the emulator. In any case, if you're allowing Rich Presence editing, why not allow leaderboard editing? We'd just want to restrict leaderboard entry submissions.

**Reply by televandalist** [reply]
> I'm hesitant to include Code Notes in the Initial Development phase. The memory map may still be in flux at that point.

We could add a disclaimer stating that the addresses could change or just not consider a console to be in the initial development phase until the memory map is good to go.

**Reply by MrOwnership** [reply]
The Initial Development phase is not meant to be an achievement development phase at all, but more so a "testing out the core" phase to see if it's ready for a future rollout. Part of my thought on allowing code notes, by admins only, in the Initial Development phase is so that we would be able to tell if a new core update has shifted around memory addresses. I wouldn't expect an admin to go in and map out the entire game memory if we are just test core compatibility.

Allowing leaderboard creation in Rollout phase but blocking entry submission sounds better than just blocking it completely. I've updated my initial post with a `Leaderboard Submissions` row.
